src/stopping.ts

```diff
@@ -0,0 +1,245 @@
+  tolerance?: CatMap<number>;
+}
+
+export interface StopIfSEMeasurementBelowThresholdInput extends EarlyStoppingInput {
+  /** Stop if the standard error of measurement drops below this level */
+  seMeasurementThreshold: CatMap<number>;
+  /** Number of items to wait for before triggering early stopping */
+  patience?: CatMap<number>;
+  /** Tolerance for standard error of measurement drop */
+  tolerance?: CatMap<number>;
+}
+
+/**
+ * Abstract class for early stopping strategies.
+ */
+export abstract class EarlyStopping {
+  protected _earlyStop: boolean;
+  protected _nItems: CatMap<number>;
+  protected _seMeasurements: CatMap<number[]>;
+  protected _logicalOperation: 'and' | 'or';
+
+  constructor({ logicalOperation = 'or' }: EarlyStoppingInput) {
+    this._seMeasurements = {};
+    this._nItems = {};
+    this._earlyStop = false;
+
+    if (!['and', 'or'].includes(logicalOperation.toLowerCase())) {
+      throw new Error(`Invalid logical operation. Expected "and" or "or". Received "${logicalOperation}"`);
+    }
+    this._logicalOperation = logicalOperation.toLowerCase() as 'and' | 'or';
+  }
+
+  public abstract get evaluationCats(): string[];
+
+  public get earlyStop() {
+    return this._earlyStop;
+  }
+
+  public get nItems() {
+    return this._nItems;
+  }
+
+  public get seMeasurements() {
+    return this._seMeasurements;
+  }
+
+  public get logicalOperation() {
+    return this._logicalOperation;
+  }
+
+  /**
+   * Update the internal state of the early stopping strategy based on the provided cats.
+   * @param {CatMap<Cat>}cats - A map of cats to update.
+   */
+  protected _updateCats(cats: CatMap<Cat>) {
+    for (const catName in cats) {
+      const cat = cats[catName];
+      const nItems = cat.nItems;
+      const seMeasurement = cat.seMeasurement;
+
+      if (nItems > (this._nItems[catName] ?? 0)) {
+        this._nItems[catName] = nItems;
+        this._seMeasurements[catName] = [...(this._seMeasurements[catName] ?? []), seMeasurement];
+      }
+    }
+  }
+
+  /**
+   * Abstract method to be implemented by subclasses to evaluate a single stopping condition.
+   * @param {string} catToEvaluate - The name of the cat to evaluate for early stopping.
+   */
+  protected abstract _evaluateStoppingCondition(catToEvaluate: string): boolean;
+
+  /**
+   * Abstract method to be implemented by subclasses to update the early stopping strategy.
+   * @param {CatMap<Cat>} cats - A map of cats to update.
+   */
+  public update(cats: CatMap<Cat>): void {
+    this._updateCats(cats);
+
+    const conditions: boolean[] = this.evaluationCats.map((catName) => this._evaluateStoppingCondition(catName));
+
+    if (this._logicalOperation === 'and') {
+      this._earlyStop = conditions.every(Boolean);
+    } else {
+      this._earlyStop = conditions.some(Boolean);
+    }
+  }
+}
+
+/**
+ * Class implementing early stopping based on a plateau in standard error of measurement.
```

Comment:
I don't get it, do you mean we have repetitive classes?

Reply:
This one stops if the SEMeasurement plateaus, meaning that is fails to decrease for a certain number of trials (called `patience`). E.g., if you have

| trial number | SEMeasurement |
| :------------ | :---------------- |
| 1 | 1.0 |
| 2 | 0.8 |
| 3 | 0.6 |
| 4 | 0.4 |
| 5 | 0.4 |
| 6 | 0.4 |
| 7 | 0.4 |

If the `patience` was set to 3, then the early stopping would have been triggered on trial 6 because the SE had failed to decrease for 3 consecutive trials.

Reply:
got it! thank you!